Move the percolator from core to its own module

martijnvg opened the pull request:
The most significant changes I had to make was moving `AbstractQueryTestCase` to the test framework module and add support the register custom plugins. Moving this base test class to the test framework was required because the `percolate` query is the first query outside of the core module and its unit test `PercolateQueryBuilderTests` extends from this test base class. The `AbstractQueryTestCase` didn't have support the register queries outside of core, so added support to the `AbstractQueryTestCase` class to register a plugin which then register its custom queries if the plugin contains that. I'm not super happy with the way needed to expose this, the entire initialization happens in [a static method](https://github.com/martijnvg/elasticsearch/blob/percolator_module/test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java#L207), so tests that want to test custom query builders now need to have a [static initializer](https://github.com/martijnvg/elasticsearch/blob/percolator_module/modules/percolator/src/test/java/org/elasticsearch/percolator/PercolateQueryBuilderTests.java#L54) that registers a plugin. If someone has a simpler/cleaner solution for this then that would be great.

Other changes:
* Lift the restriction that only one percolator could be added per index. This validation existed in MapperService, but because the percolator moved to a module it could no longer exist there. Instead of bringing it back it was removed. This validation existed since the percolator cache only supported one percolator query per document, since the percolator cache has been removed this restriction could removed as well.
* While moving percolator tests to the new module, also removed a couple of tests for the deprecated percolate and mpercolate api. These APIs are now sugar APIs for bwc support and redirect to the search and msearch APIs. Some tests were still testing as if percolate and mpercolate API did the percolation, but this no longer the case and these tests could be removed.

Note: The breaking part is that the percolate methods have been removed from the `Client` class in the Java api.

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -283,7 +298,11 @@ protected void configure() {
+        pluginsService.processModules(modulesBuilder);
+        injector = modulesBuilder.createInjector();
+        IndexScopedSettings indexScopedSettings = injector.getInstance(IndexScopedSettings.class);
+        idxSettings = IndexSettingsModule.newIndexSettings(index, indexSettings, indexScopedSettings);
```
why is this moved down here and why do you need to Initialize `idxSettings` so late?

martijnvg replied:
I need it after the guice injector has been created, so that I can get `IndexScopedSettings` instance. The percolator query builder test that tests the `doToQuery` would otherwise fail, since there an index setting is checked that the `PercolatorPlugin` registers.

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -330,6 +349,10 @@ public void onRemoval(ShardId shardId, Accountable accountable) {
         namedWriteableRegistry = injector.getInstance(NamedWriteableRegistry.class);
     }
 
+    protected static void registerPlugin(Class<? extends Plugin> pluginClass) {
```
instead of doing this, can we try to make `init()` a member method and see what's the impact? I mean if it's cheap why having the hassle to manage static deps?

martijnvg replied:
I'll try out and see what the impact is. Before opening I was doubting if this should be done as I was unsure of the impact. But if it works out, this will end up being much cleaner.

s1monw replied:
I just spoke to @cbuescher and it has a perf impact. I think we should move that entire thing into a Holder class that we pull in and only create once and put it in a static member ie. something like this:

```Java

@Before
public void setUp() {
  super.setUp();
  if (holderReference == null) { // holderReference is static
    holderReference = new Holder(plugins);
  }
}

@AfterClass
public void afterClass() {
  holderReference = null;
}
```

martijnvg replied:
I see, I'll try that out.

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/IndexSettingsModule.java:
```diff
@@ -66,6 +66,16 @@ public static IndexSettings newIndexSettings(Index index, Settings settings, Set
         return new IndexSettings(metaData, Settings.EMPTY, (idx) -> Regex.simpleMatch(idx, metaData.getIndex().getName()), new IndexScopedSettings(Settings.EMPTY, settingSet));
     }
 
+    public static IndexSettings newIndexSettings(Index index, Settings settings, IndexScopedSettings indexScopedSettings) {
```
I really wonder why we need this.

martijnvg replied:
I added this method, because I inject a guice managed `IndexScopedSettings` instance in `AbstractQueryTestCase#init()`

martijnvg commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -374,11 +224,189 @@ public IndexFieldDataService fieldData() {
     }
 
-    protected final QB createTestQueryBuilder() {
+    static class Holder {
```
I couldn't come up with a better name...

s1monw replied:
ServiceHolder?

martijnvg commented:
@s1monw @cbuescher I've changed the initialization method to be a member method and introduced this the `Holder` clause that gets initialized only once per test suite.

nik9000 commented on core/src/main/java/org/elasticsearch/client/Client.java:
```diff
@@ -420,36 +414,6 @@
     /**
-     * Percolates a request returning the matches documents.
-     */
-    ActionFuture<PercolateResponse> percolate(PercolateRequest request);
```
One thing that has come up with reindex being in a plugin: users of the Java API have trouble finding it. I don't know what, if anything, we should do about that. This will have the same problem.

martijnvg replied:
that is true. The docs that described how to integrate with percolate api via the Java client have been replaced with: https://www.elastic.co/guide/en/elasticsearch/client/java-api/master/java-specialized-queries.html#java-query-percolate-query

I'll bring back the percolator java api docs with a description on how to access the api now that the percolator api has been removed from the `Client` interface. Otherwise people that wish to use the percolator api a bit longer are in the dark before cutting over to the search api.

nik9000 commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -171,7 +171,7 @@ public boolean isFilter() {
     }
 
-    void setIsFilter(boolean isFilter) {
+    public void setIsFilter(boolean isFilter) {
```
If public maybe should have javadoc?

nik9000 commented on core/src/test/java/org/elasticsearch/ExceptionSerializationTests.java:
```diff
@@ -743,7 +743,6 @@ public void testIds() {
         ids.put(105, org.elasticsearch.cluster.routing.RoutingException.class);
         ids.put(106, org.elasticsearch.index.shard.IndexShardRecoveryException.class);
         ids.put(107, org.elasticsearch.repositories.RepositoryMissingException.class);
-        ids.put(108, org.elasticsearch.index.percolator.PercolatorException.class);
```
Maybe put null in the array like with do with other exceptions?

martijnvg replied:
good point

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -149,209 +158,50 @@
-        mapperService = null;
-        bitsetFilterCache = null;
-        scriptService = null;
+        holder.close();
```
this will cause an NPE if you fail to initialize maybe use `IOUtils.close(holder)`

martijnvg replied:
good point

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -149,209 +158,50 @@
 
-    private static Index index;
-
     protected static Index getIndex() {
```
we can now make all these instance methods no?

martijnvg replied:
I think so too.

s1monw commented:
left some minors LGTM otherwise 👍
